### DISCUSSION TITLE
Modelling: Tracking state of a resource

### DISCUSSION THREAD

**Discussion by guija** [TimefoldAI/timefold-solver#511] [General]
Hi all, first of all, thanks for your framework! It's great and I already used it in the past successfully. Now I have a different modelling problem:

- Have a set of resources (e.g. a battery with a charging state, e.g. between 0% and 100%)
- I have a list of activities (which are a window with start and end date). Each activities has some influence on the resource (e.g. discharge battery by x percent or charge battery by x%)

Multiple activities can run in parallel and have influence on the same resource (e.g. same instance of a battery).

Also there should be a constraint that says: The battery charge should never be under 0% (e.g. if it is zero at point in time z then afterwards only activities can be executed that charge the battery).

I thought to use one or a combination of the following approaches:

- Chained variables: This seems unfeasible as you can only form a chain without having multiple successors or predecessors. So having multiple activities that influence the same resource cannot be done.
- Shadow variables: Using a shadow variable that determines the charging state of the battery. Having a condition that finds all the activities that are connected (assigned to the battery) and overlap in time and check whether their sum at a specific point in time is >= 0. However there is no connection to the previous activities or the previous state of the battery charge.
- **Using the time grain pattern: Chaining the grains with grains having a shadow variable that is inferring the battery charge based on the previous grain and by checking (=having constraints) which activities are assigned to this grain and how these activities perform on the battery charge.**

So my question is:
- Do you think in general that it is possible or even sensible to model this in timefold?
- If yes, does someone have an idea of how to model this or is any of the ideas mentioned above any good?

I will try testing out the last approach for now but I would be grateful for ideas.

**Comment by thimmwork** [comment]
Can a list of parallel running activities use up battery faster than it can recharge? Is charging subject to time constraints?
If not, a greedy algorithm might tell you as well when to charge.

Is it safe to assume that the tasks will spread their effect on the battery evenly/linearly across the time grains?
(e.g. would a process that takes 1 hour and use 8% battery use 2% for each 15 minute grain? and will it use 8% independent of the charging state?)
If yes, it looks doable to me. Otherwise the complexity might quickly outrun the benefit.
